src/transformers/models/canine/tokenization_canine.py

```diff
@@ -0,0 +1,239 @@
+        **kwargs
+    ):
+        bos_token = AddedToken(bos_token, lstrip=False, rstrip=False) if isinstance(bos_token, str) else bos_token
+        eos_token = AddedToken(eos_token, lstrip=False, rstrip=False) if isinstance(eos_token, str) else eos_token
+        sep_token = AddedToken(sep_token, lstrip=False, rstrip=False) if isinstance(sep_token, str) else sep_token
+        cls_token = AddedToken(cls_token, lstrip=False, rstrip=False) if isinstance(cls_token, str) else cls_token
+        pad_token = AddedToken(pad_token, lstrip=False, rstrip=False) if isinstance(pad_token, str) else pad_token
+
+        # Mask token behave like a normal word, i.e. include the space before it
+        mask_token = AddedToken(mask_token, lstrip=True, rstrip=False) if isinstance(mask_token, str) else mask_token
+
+        super().__init__(
+            bos_token=bos_token,
+            eos_token=eos_token,
+            sep_token=sep_token,
+            cls_token=cls_token,
+            pad_token=pad_token,
+            mask_token=mask_token,
+            add_prefix_space=add_prefix_space,
+            model_max_length=model_max_length,
+            **kwargs,
+        )
+
+        # Creates a mapping for looking up the IDs of special symbols.
+        self._special_codepoints: Dict[Text, int] = {}
+        for codepoint, name in SPECIAL_CODEPOINTS.items():
+            self._special_codepoints[name] = codepoint
+
+        # Creates a mapping for looking up the string forms of special symbol IDs.
+        self._special_codepoint_strings: Dict[int, Text] = {
+            codepoint: name for name, codepoint in self._special_codepoints.items()
+        }
+
+        self._unicode_vocab_size = 1114112  # Unicode defines 1,114,112 total “codepoints”
+        self._num_special_tokens = len(self._special_codepoints)
+
+    @property
+    def vocab_size(self):
+        return self._unicode_vocab_size + self._num_special_tokens
+
+    def _tokenize(self, text: str) -> List[str]:
+        """Tokenize a string (i.e. perform character splitting)."""
+
+        pieces = [x for x in text]
+
+        return pieces
+
+    def _convert_token_to_id(self, token: str):
+        """Converts a token (i.e. a Unicode character) in an id (i.e. its integer Unicode code point value)."""
+
+        if token in self._special_codepoints:
+            return self._special_codepoints[token]
+        try:
+            return ord(token)
+        except TypeError:
+            raise ValueError(f"invalid token: '{token}'")
+
+    def _convert_id_to_token(self, index: int):
+        """Converts an integer Unicode code point (integer) in a token (str)."""
+
+        if index in self._special_codepoint_strings:
+            return self._special_codepoint_strings[index]
+        try:
+            return chr(index)
+        except TypeError:
+            raise ValueError(f"invalid id: {index}")
+
+    def convert_tokens_to_string(self, tokens):
+        out_string = "".join(tokens).strip()
```

Comment:
Just return it. The name `out_string` isn't providing much value to the reader. Also I wouldn't do `.strip()` since presumably it's better to handle exactly what is in the input.